docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,296 @@
+
+- To create a mapping for dynamic enclave memory allocation, mmap must be called with an open enclave file descriptor and with PROT_READ | PROT_WRITE for protection flags. 
+  - Enclave must issue EACCEPT for the pages after mmap before it can modify the content of the pages and extend/reduce permissions in secure way.
+- The offset in mmap parameter must be zero for enclaves.
+- MAP_* flags must be MAP_SHARED  | MAP_FIXED masked with optional flags:
+    - MAP_POPULATE: hint for kernel to EAUG pages as soon as possible.  
+    - MAP_GROWSDOWN: used for stacks.  The mapping will grow down to the next mapping.
+- If and only if the address range are within the ELRANGE of the enclave associated with the file descriptor, the mapping will be created. However, user space should not expect EAUG be done by the mmap call.
+  - The kernel can choose EAUG pages immediately (likely for MAP_POPULATE), or EAUG pages upon page faults within the VMA, similar to how kernel would allocate regular memory.
+- The kernel will assume the newly requested mapping is for dynamic allocation and initial permissions must be RW until user space request changes later. 
+
+**Implementation Notes:** Current [SGX kernel patches](https://patchwork.kernel.org/project/intel-sgx/patch/20201112220135.165028-11-jarkko@kernel.org/) limit PTE permissions to the EPCM permissions given in SEC_INFO during EADD IOCTL calls. The dynamic allocation mappings should not be subject to those limits. A possible implementation may have these changes:
+  - sgx_encl_may_map 
+    - enforces RW permissions for pages other than those loaded due to EADD or ECREATE.
+    - set up flags to track dynamic pages: type, permissions flag
+  - sgx_vma_mprotect
+    - Allow permissions changes to dynamic pages within limitations of OS policies, e.g.,
+	    - never allow WX
+	    - SELinux policy specific to SGX enclaves
+    - update flags for the dynamic pages
+
+## munmap 
+Calling munmap on an enclave page (dynamic allocated or not) has exactly the same effect of calling munmap on a regular RAM page. No sgx specific interface is needed. No behavior changes to current kernel space implementation.
+
+### Remarks
+
+- Enclave memory mapings are shared (MAP_SHARED). The mappings in shared processes are kept alive and independently until the process exits
+    - munmap and closing file descriptors are not required for user space. A dead process automatically releases all mappings and file descriptors. 
+- Upon all enclave mappings are removed and file handles to the enclave are closed, either by explicit munmap/fclose syscalls or when all hosting apps exited:
+    - The kernel may mark its remaining pages are reclaimable and issue EREMOVE on them any time the kernel deems appropriate.
+
+## mprotect IOCTL
+This IOCTL emulates the mprotect syscall with SGX specific extensions. In future kernel implementation, it could be mprotect or pkey_mprotect syscall with sgx extensions for the "prot" parameter.
```

Comment:
```suggestion
This IOCTL emulates the mprotect syscall with SGX specific extensions. In a future kernel implementation, it could be mprotect or pkey_mprotect syscall with SGX extensions for the "prot" parameter.
```